components/board/genericlinux/gpio.go

```diff
@@ -312,11 +317,18 @@ func (pin *gpioPin) halfPwmCycle(ctx context.Context, shouldBeOn bool) bool {
 
 func (pin *gpioPin) softwarePwmLoop(ctx context.Context) {
 	for {
-		if !pin.halfPwmCycle(ctx, true) {
+		select {
+		case <-ctx.Done():
 			return
```

Comment:
Note to myself: I was going to add in a second select statement that only checked if the context was canceled, so that we can't mistakenly keep going if the context is canceled but we also get the signal that the loop should continue. However, if we do continue past here, we're going to get into `halfPwmCycle` and immediately notice that the context is canceled and it's time to shut everything down anyway. So, no need for a redundant check here.